### PR TITLE
UI Component to convert clusters into abstract syntax tree

### LINKED ISSUE #418: Grasshopper_Toolkit: Convert from GH Cluster to abstract syntax tree
#### Description:
Limit this capability to BHoM components unless trivial to include GH components

### PR DESCRIPTION
**Pull request by adecler**
### NOTE: Depends on 
- https://github.com/BHoM/BHoM/pull/682
- https://github.com/BHoM/BHoM_Engine/pull/1451
   
### Issues addressed by this PR

Closes #418 

### Test files
See https://github.com/BHoM/Node2Code_Toolkit/pull/2 for more details

### Additional comments
- Some of the code should move to BHoM_UI but I kept everything in Grasshopper for now to minimise the footprint (since this was developed over a long period of time)

### CODE REVIEW

File: Grasshopper_Engine/Convert/ToNode.cs [hunk 1]
```diff
@@ -0,0 +1,261 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2018, the respective contributors. All rights reserved.
```
Comment (FraserGreenroyd) [inline Grasshopper_Engine/Convert/ToNode.cs]: Copyright should be 2020 (sorry!) - though after discussion with @al-fisher earlier today we're going to get BHoM bot to do a mass update on copyright, so feel free to not change if you're happy with letting BHoM Bot do it on this PR :smile:

File: Grasshopper_Engine/Grasshopper_Engine.csproj [hunk 2]
```diff
@@ -66,6 +70,15 @@
       <HintPath>..\packages\Grasshopper.0.9.76\lib\net35\Grasshopper.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="Node2Code_Engine">
+      <HintPath>..\..\Node2Code_Toolkit\Build\Node2Code_Engine.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Node2Code_oM, Version=3.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (FraserGreenroyd) [inline Grasshopper_Engine/Grasshopper_Engine.csproj]: Reference shouldn't include version, culture, or processor architecture

File: Grasshopper_Engine/Convert/ToNode.cs (hunk 1 above)
Comment (al-fisher) [reply Grasshopper_Engine/Convert/ToNode.cs]: Yep @FraserGreenroyd I would stand down any humans manually updating copyright statements now. @BHoMBot can take over this milestone.
As discussed happy/not end of the world if takes a week or so with 2019 copyrights merged in meantime.
@adecler

**Comment by al-fisher** [comment]
@adecler just popping a `no-not-merge` on this PR for the moment. 
We will need to update installers and CI dependencies in one of two ways before happily merging:
1. We can either update the dependencies for Grasshopper_Toolkit to refer also to Node2Code, or
1. we can consider (now with Node2Code public) moving Node2Code_oM/Engine into core BHoM as their own projects. Could be inside BHoM_UI or inside oM and Engine respectively. 
Liking the latter actually on first thinking - but good to share thoughts. 

I will pop a catch up call in for Friday as usual and we can discuss 
Cheers!

**Comment by adecler** [comment]
the dependency to Node2Code has now been removed. So ready to be merged if you guys are happy.

**Comment by adecler** [comment]
@FraserGreenroyd , it appears that AppVeyor is not supporting C#7 language at the moment. Not a big deal for now but it might be good to keep in mind if you have other weird errors from CI.

**Comment by IsakNaslundBh** [comment]
> @FraserGreenroyd , it appears that AppVeyor is not supporting C#7 language at the moment. Not a big deal for now but it might be good to keep in mind if you have other weird errors from CI.

Think this might be intentional as a lot of people, including myself, are using VS15 which has no support for C#7, hence merging C#7 code in would make it impossible for us to compile, so think that check is a good one for now. But will let @FraserGreenroyd answer as well.

**Comment by al-fisher** [comment]
Yes - we are intentionally configuring the environment set ups for CI as behind the latest versions of VS etc. For the reasons above.

**Comment by adecler** [comment]
Makes sense. I think I vaguely remember a conversation around that but though I would put a reminder comment here just in case. Looks like I was the only one with a bad memory 😋

**Comment by FraserGreenroyd** [comment]
Seems like everyone knows the answer so I'll just comment to confirm the answers above :smile:

**Pull request merged**